homeassistant/components/mqtt/discovery.py

```diff
@@ -141,6 +143,34 @@ async def async_start(  # noqa: C901
     """Start MQTT Discovery."""
     mqtt_data = get_mqtt_data(hass)
     mqtt_integrations = {}
+    platform_setup_lock = asyncio.Lock()
+
+    async def _async_jit_component_setup(
+        discovery_payload: MQTTDiscoveryPayload,
+    ) -> None:
+        """Perform just in time components set up."""
+        discovery_hash = discovery_payload.discovery_data[ATTR_DISCOVERY_HASH]
+        component, discovery_id = discovery_hash
+        with suppress(ValueError):
+            component = Platform(component)
+        async with platform_setup_lock:
+            if component not in mqtt_data.platforms_loaded:
+                await async_forward_entry_setup_and_setup_discovery(
+                    hass, config_entry, {component}
+                )
```

Comment:
If its ok to setup multiple platforms at the same time, it might be faster to only hold the lock to add the platform to `platforms_loaded` and than await the forward outside the lock

Reply:
The platform dispatcher is added during platform setup, I want to be sure platforms_loaded is not set before the dispatcher listening.

https://github.com/home-assistant/core/blob/mqtt-jit-platform-setup/homeassistant/components/mqtt/mixins.py#L317-L380

Reply:
Is it possible that
`switch` gets discovered while `light` is being setup by `async_forward_entry_setup_and_setup_discovery` so it has to wait for the `light` platform to finish so it can get the lock?

Reply:
That is possible

Reply:
I could add a test that simulates a multiple platform discovery

Reply:
test added

Reply:
Should the lock be per component? 

I was concerned that if you get a discovery for a `light` and than `switch` and than a `vacuum` I think everything ends up loading in series instead of all the lights waiting for the light platform, all the switches waiting for the switch platform, etc

Reply:
Done